File: mobile/iosApp/iosApp/ViewModels/DashboardViewModel.swift
// SECT Mobile — iOS Dashboard ViewModel (RBAC: Enseignant/Étudiant)
import SwiftUI
import Shared

@MainActor
class DashboardViewModel: ObservableObject {
    @Published var user: User? = nil
    
    // Stats enrichies selon le rôle
    @Published var enseignantStats: EnseignantStats? = nil
    @Published var etudiantStats: EtudiantStats? = nil
    
    // Legacy (pour compatibilité)
    @Published var upcomingEpreuves: [Epreuve] = []
    @Published var totalEpreuves: Int = 0
    @Published var enCours: Int = 0
    @Published var planifiees: Int = 0
    
    @Published var isLoading = false
    @Published var error: String? = nil
    
    private let repository = KoinRepositoryProvider.shared.repository
    
    var isEnseignant: Bool {
        user?.role == .enseignant
    }
    
    var isEtudiant: Bool {
        user?.role == .etudiant
    }
    
    func loadDashboard() async {
        isLoading = true
        error = nil
        do {
            // Load current user profile
            user = try await repository.getCurrentUser()
            
            // Charger les stats selon le rôle
            if isEnseignant {
                try await loadEnseignantStats()
            } else if isEtudiant {
                try await loadEtudiantStats()
            }
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }
    
    // MARK: - Stats Enseignant
    
    private func loadEnseignantStats() async throws {
        let stats = try await repository.getStatsEnseignant()
        enseignantStats = stats
        
        // Compatibilité legacy
        totalEpreuves = Int(stats.nbEpreuves)
        enCours = Int(stats.nbEpreuvesActives)
        planifiees = stats.epreuvesAVenir.count
        
        upcomingEpreuves = stats.epreuvesAVenir.map { epreuve in
            Epreuve(
                id: epreuve.id,
                enseignantId: "",
                titre: epreuve.titre,
                description: "",
                duree: epreuve.duree,
                dateDebut: epreuve.date,
                dateFin: epreuve.dateFin,
                melangeQuestions: false,
                melangePropositions: false,
                blocageRetour: false,
                statut: .planifiee,
                filiereId: "",
                uniteEnseignementId: nil,
                niveau: nil,
                sessionExamen: .normale,
                anneeAcademiqueId: nil,
                createdAt: "",
                updatedAt: "",
                deletedAt: nil,
                proctoringActif: false,
                verificationIdentite: false,
                generationMode: .manuelle,
                isTemplate: false,
                noteTotal: 20.0,
                clotureeAt: nil,
                clotureeAutomatiquement: false,
                raisonCloture: nil,
                delaiGrace: 0,
                epreuveOrigineId: nil,
                questionCount: KotlinInt(int: 0),
                totalPoints: KotlinDouble(double: 0.0),
                enseignant: nil,
                filiere: nil,
                questions: nil
            )
        }
    }
    
    // MARK: - Stats Étudiant
    
    private func loadEtudiantStats() async throws {
        let stats = try await repository.getStatsEtudiant()
        etudiantStats = stats
        
        // Compatibilité legacy
        totalEpreuves = Int(stats.nbEpreuvesTerminees + stats.nbEpreuvesAVenir)
        enCours = stats.sessionEnCours != nil ? 1 : 0
        planifiees = Int(stats.nbEpreuvesAVenir)
        
        upcomingEpreuves = stats.epreuvesAVenir.map { epreuve in
            Epreuve(
                id: epreuve.id,
                enseignantId: "",
                titre: epreuve.titre,
                description: "",
                duree: epreuve.duree,
                dateDebut: epreuve.date,
                dateFin: epreuve.dateFin,
                melangeQuestions: false,
                melangePropositions: false,
                blocageRetour: false,
                statut: .planifiee,
                filiereId: "",
                uniteEnseignementId: nil,
                niveau: nil,
                sessionExamen: .normale,
                anneeAcademiqueId: nil,
                createdAt: "",
                updatedAt: "",
                deletedAt: nil,
                proctoringActif: false,
                verificationIdentite: false,
                generationMode: .manuelle,
                isTemplate: false,
                noteTotal: 20.0,
                clotureeAt: nil,
                clotureeAutomatiquement: false,
                raisonCloture: nil,
                delaiGrace: 0,
                epreuveOrigineId: nil,
                questionCount: KotlinInt(int: epreuve.nbQuestions),
                totalPoints: KotlinDouble(double: epreuve.totalPoints),
                enseignant: nil,
                filiere: nil,
                questions: nil
            )
        }
    }
    
    func refresh() async {
        await loadDashboard()
    }
}

File: mobile/androidApp/src/main/kotlin/com/sect/mobile/android/ui/viewmodel/DashboardViewModel.kt
// SECT Mobile — DashboardViewModel (stats, raccourcis, épreuves à venir)
// SECT-MOBILE-FOCUS : adaptatif enseignant / étudiant avec statistiques complètes
package com.sect.mobile.android.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sect.mobile.shared.domain.model.*
import com.sect.mobile.shared.domain.enum.Role
import com.sect.mobile.shared.domain.repository.SECTRepositoryInterface
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class DashboardViewModel(private val repository: SECTRepositoryInterface) : ViewModel() {

    private val _user = MutableStateFlow<User?>(null)
    val user: StateFlow<User?> = _user.asStateFlow()

    // Stats enrichies selon le rôle
    private val _enseignantStats = MutableStateFlow<UiState<EnseignantStats>>(UiState.Loading)
    val enseignantStats: StateFlow<UiState<EnseignantStats>> = _enseignantStats.asStateFlow()

    private val _etudiantStats = MutableStateFlow<UiState<EtudiantStats>>(UiState.Loading)
    val etudiantStats: StateFlow<UiState<EtudiantStats>> = _etudiantStats.asStateFlow()

    private val _upcomingEpreuves = MutableStateFlow<UiState<List<Epreuve>>>(UiState.Loading)
    val upcomingEpreuves: StateFlow<UiState<List<Epreuve>>> = _upcomingEpreuves.asStateFlow()

    val isEnseignant: Boolean
        get() = _user.value?.role == Role.ENSEIGNANT

    val isEtudiant: Boolean
        get() = _user.value?.role == Role.ETUDIANT

    init {
        loadDashboard()
    }

    fun loadDashboard() {
        viewModelScope.launch {
            try {
                _user.value = repository.getCurrentUser()
            } catch (_: Exception) { }

            val role = _user.value?.role

            when (role) {
                Role.ENSEIGNANT -> loadEnseignantDashboard()
                Role.ETUDIANT -> loadEtudiantDashboard()
                else -> { /* ADMIN/RESPONSABLE ne devrait pas être ici */ }
            }
        }
    }

    /**
     * Dashboard enseignant :
     * - Stats complètes (documents, questions, épreuves, corrections en attente)
     * - Épreuves récentes avec performances
     * - Corrections en attente
     * - Épreuves à venir
     */
    private suspend fun loadEnseignantDashboard() {
        coroutineScope {
            launch {
                try {
                    val stats = repository.getStatsEnseignant()
                    _enseignantStats.value = UiState.Success(stats)
                    
                    // Charger aussi les épreuves à venir pour compatibilité
                    val epreuvesAVenir = stats.epreuvesAVenir.map { epreuve ->
                        Epreuve(
                            id = epreuve.id,
                            titre = epreuve.titre,
                            description = "",
                            statut = com.sect.mobile.shared.domain.enum.StatutEpreuve.valueOf(epreuve.statut),
                            dateDebut = epreuve.date,
                            dateFin = epreuve.dateFin,
                            duree = epreuve.duree,
                            questionCount = 0,
                            totalPoints = 0.0,
                            filiereId = "",
                            enseignantId = "",
                            createdAt = "",
                            updatedAt = "",
                            melangeQuestions = false,
                            melangePropositions = false,
                            blocageRetour = false,
                            sessionExamen = com.sect.mobile.shared.domain.enum.SessionExamen.NORMALE,
                            generationMode = com.sect.mobile.shared.domain.enum.ModeGeneration.MANUELLE
                        )
                    }
                    _upcomingEpreuves.value = UiState.Success(epreuvesAVenir)
                } catch (e: Exception) {
                    _enseignantStats.value = UiState.Error(e.message ?: "Erreur")
                    _upcomingEpreuves.value = UiState.Error(e.message ?: "Erreur")
                }
            }
        }
    }

    /**
     * Dashboard étudiant :
     * - Stats complètes (moyenne, meilleure note, évolution scores)
     * - Épreuves à venir avec détails
     * - Résultats récents
     * - Session en cours (si applicable)
     */
    private suspend fun loadEtudiantDashboard() {
        coroutineScope {
            launch {
                try {
                    val stats = repository.getStatsEtudiant()
                    _etudiantStats.value = UiState.Success(stats)
                    
                    // Charger aussi les épreuves à venir pour compatibilité
                    // Note : EpreuveAVenirEtudiant n'expose pas `statut` → défaut PLANIFIEE (épreuve à venir)
                    val epreuvesAVenir = stats.epreuvesAVenir.map { epreuve ->
                        Epreuve(
                            id = epreuve.id,
                            titre = epreuve.titre,
                            description = "",
                            statut = com.sect.mobile.shared.domain.enum.StatutEpreuve.PLANIFIEE,
                            dateDebut = epreuve.date,
                            dateFin = epreuve.dateFin,
                            duree = epreuve.duree,
                            questionCount = epreuve.nbQuestions,
                            totalPoints = epreuve.totalPoints,
                            filiereId = "",
                            enseignantId = "",
                            createdAt = "",
                            updatedAt = "",
                            melangeQuestions = false,
                            melangePropositions = false,
                            blocageRetour = false,
                            sessionExamen = com.sect.mobile.shared.domain.enum.SessionExamen.NORMALE,
                            generationMode = com.sect.mobile.shared.domain.enum.ModeGeneration.MANUELLE
                        )
                    }
                    _upcomingEpreuves.value = UiState.Success(epreuvesAVenir)
                } catch (e: Exception) {
                    _etudiantStats.value = UiState.Error(e.message ?: "Erreur")
                    _upcomingEpreuves.value = UiState.Error(e.message ?: "Erreur")
                }
            }
        }
    }

    fun refresh() = loadDashboard()
}
